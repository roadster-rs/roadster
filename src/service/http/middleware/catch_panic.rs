use crate::app::context::AppContext;
use crate::error::RoadsterResult;
use crate::service::http::middleware::Middleware;
use axum::Router;
use serde_derive::{Deserialize, Serialize};
use tower_http::catch_panic::CatchPanicLayer;

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case", default)]
pub struct CatchPanicConfig {}

pub struct CatchPanicMiddleware;
impl<S: Send + Sync + 'static> Middleware<S> for CatchPanicMiddleware {
    fn name(&self) -> String {
        "catch-panic".to_string()
    }

    fn enabled(&self, context: &AppContext<S>) -> bool {
        context
            .config()
            .service
            .http
            .custom
            .middleware
            .catch_panic
            .common
            .enabled(context)
    }

    fn priority(&self, context: &AppContext<S>) -> i32 {
        context
            .config()
            .service
            .http
            .custom
            .middleware
            .catch_panic
            .common
            .priority
    }

    fn install(&self, router: Router, _context: &AppContext<S>) -> RoadsterResult<Router> {
        let router = router.layer(CatchPanicLayer::new());

        Ok(router)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::config::app_config::AppConfig;
    use rstest::rstest;

    #[rstest]
    #[case(false, Some(true), true)]
    #[case(false, Some(false), false)]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn enabled(
        #[case] default_enable: bool,
        #[case] enable: Option<bool>,
        #[case] expected_enabled: bool,
    ) {
        // Arrange
        let mut config = AppConfig::test(None).unwrap();
        config.service.http.custom.middleware.default_enable = default_enable;
        config
            .service
            .http
            .custom
            .middleware
            .catch_panic
            .common
            .enable = enable;

        let context = AppContext::<()>::test(Some(config), None, None).unwrap();

        let middleware = CatchPanicMiddleware;

        // Act/Assert
        assert_eq!(middleware.enabled(&context), expected_enabled);
    }

    #[rstest]
    #[case(None, 0)]
    #[case(Some(1234), 1234)]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn priority(#[case] override_priority: Option<i32>, #[case] expected_priority: i32) {
        // Arrange
        let mut config = AppConfig::test(None).unwrap();
        if let Some(priority) = override_priority {
            config
                .service
                .http
                .custom
                .middleware
                .catch_panic
                .common
                .priority = priority;
        }

        let context = AppContext::<()>::test(Some(config), None, None).unwrap();

        let middleware = CatchPanicMiddleware;

        // Act/Assert
        assert_eq!(middleware.priority(&context), expected_priority);
    }
}
