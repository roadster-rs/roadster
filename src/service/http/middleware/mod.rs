pub mod catch_panic;
pub mod compression;
pub mod default;
pub mod request_id;
pub mod sensitive_headers;
pub mod size_limit;
pub mod timeout;
pub mod tracing;

#[mockall_double::double]
use crate::app_context::AppContext;
use axum::Router;

/// Allows initializing and installing middleware on the app's [Router]. The type `S` is the
/// custom [crate::app::App::State] defined for the app.
///
/// This trait is provided in addition to [crate::service::http::initializer::Initializer] because installing
/// middleware is a bit of a special case compared to a general initializer:
///     1. The order in which middleware runs matters. For example, we want
///        [tower_http::sensitive_headers::SetSensitiveRequestHeadersLayer] to run before
///        [tower_http::trace::TraceLayer] to avoid logging sensitive headers.
///     2. Because of how axum's [Router::layer] method installs middleware, the order in which
///        middleware is installed is the reverse of the order it will run when handling a request.
///        Therefore, we install the middleware in the reverse order that we want it to run (this
///        is done automatically by Roadster based on [Middleware::priority]).
pub trait Middleware<S>: Send {
    fn name(&self) -> String;
    fn enabled(&self, context: &AppContext<S>) -> bool;
    /// Used to determine the order in which the middleware will run when handling a request. Smaller
    /// numbers will run before larger numbers. For example, a middleware with priority `-10`
    /// will run before a middleware with priority `10`.
    ///
    /// If two middlewares have the same priority, they are not guaranteed to run or be installed
    /// in any particular order relative to each other. This may be fine for many middlewares.
    ///
    /// If the order in which your middleware runs doesn't particularly matter, it's generally
    /// safe to set its priority as `0`.
    ///
    /// Note: Because of how axum's [Router::layer] method installs middleware, the order in which
    /// middleware is installed is the reverse of the order it will run when handling a request.
    /// Therefore, we install the middleware in the reverse order that we want it to run (this
    /// is done automatically by Roadster based on [Middleware::priority]). So, a middleware
    /// with priority `-10` will be _installed after_ a middleware with priority `10`, which will
    /// allow the middleware with priority `-10` to _run before_ a middleware with priority `10`.
    fn priority(&self, context: &AppContext<S>) -> i32;
    fn install(&self, router: Router, context: &AppContext<S>) -> anyhow::Result<Router>;
}
