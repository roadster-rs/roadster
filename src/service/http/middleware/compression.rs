use crate::app_context::AppContext;
use crate::service::http::middleware::Middleware;
use axum::Router;
use serde_derive::{Deserialize, Serialize};

use tower_http::compression::CompressionLayer;
use tower_http::decompression::RequestDecompressionLayer;

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case", default)]
pub struct ResponseCompressionConfig {}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case", default)]
pub struct RequestDecompressionConfig {}

pub struct ResponseCompressionMiddleware;
impl<S> Middleware<S> for ResponseCompressionMiddleware {
    fn name(&self) -> String {
        "response-compression".to_string()
    }

    fn enabled(&self, context: &AppContext, _state: &S) -> bool {
        context
            .config()
            .service
            .http
            .custom
            .middleware
            .response_compression
            .common
            .enabled(context)
    }

    fn priority(&self, context: &AppContext, _state: &S) -> i32 {
        context
            .config()
            .service
            .http
            .custom
            .middleware
            .response_compression
            .common
            .priority
    }

    fn install(&self, router: Router, _context: &AppContext, _state: &S) -> anyhow::Result<Router> {
        let router = router.layer(CompressionLayer::new());

        Ok(router)
    }
}

pub struct RequestDecompressionMiddleware;
impl<S> Middleware<S> for RequestDecompressionMiddleware {
    fn name(&self) -> String {
        "request-decompression".to_string()
    }

    fn enabled(&self, context: &AppContext, _state: &S) -> bool {
        context
            .config()
            .service
            .http
            .custom
            .middleware
            .request_decompression
            .common
            .enabled(context)
    }

    fn priority(&self, context: &AppContext, _state: &S) -> i32 {
        context
            .config()
            .service
            .http
            .custom
            .middleware
            .request_decompression
            .common
            .priority
    }

    fn install(&self, router: Router, _context: &AppContext, _state: &S) -> anyhow::Result<Router> {
        let router = router.layer(RequestDecompressionLayer::new());

        Ok(router)
    }
}
