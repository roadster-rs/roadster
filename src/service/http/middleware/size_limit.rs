use crate::app::context::AppContext;
use crate::error::RoadsterResult;
use crate::service::http::middleware::Middleware;
use anyhow::anyhow;
use axum::Router;
use axum_core::extract::FromRef;
use byte_unit::Byte;
use byte_unit::Unit::MB;
use byte_unit::rust_decimal::prelude::ToPrimitive;
use serde_derive::{Deserialize, Serialize};
use tower_http::limit::RequestBodyLimitLayer;
use validator::Validate;

#[derive(Debug, Clone, Validate, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case", default)]
#[non_exhaustive]
pub struct SizeLimitConfig {
    pub limit: Byte,
}

impl Default for SizeLimitConfig {
    fn default() -> Self {
        Self {
            #[allow(clippy::expect_used)]
            limit: Byte::from_u64_with_unit(5, MB).expect("Unable to build Byte unit."),
        }
    }
}

pub struct RequestBodyLimitMiddleware;
impl<S> Middleware<S> for RequestBodyLimitMiddleware
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
{
    fn name(&self) -> String {
        "request-body-size-limit".to_string()
    }

    fn enabled(&self, state: &S) -> bool {
        AppContext::from_ref(state)
            .config()
            .service
            .http
            .custom
            .middleware
            .size_limit
            .common
            .enabled(state)
    }

    fn priority(&self, state: &S) -> i32 {
        AppContext::from_ref(state)
            .config()
            .service
            .http
            .custom
            .middleware
            .size_limit
            .common
            .priority
    }

    fn install(&self, router: Router, state: &S) -> RoadsterResult<Router> {
        let limit = &AppContext::from_ref(state)
            .config()
            .service
            .http
            .custom
            .middleware
            .size_limit
            .custom
            .limit
            .as_u64()
            .to_usize();

        let limit = match limit {
            Some(limit) => limit,
            None => return Err(anyhow!("Unable to convert bytes from u64 to usize").into()),
        };

        let router = router.layer(RequestBodyLimitLayer::new(*limit));

        Ok(router)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::config::AppConfig;
    use rstest::rstest;

    #[rstest]
    #[case(false, Some(true), true)]
    #[case(false, Some(false), false)]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn size_limit_enabled(
        #[case] default_enable: bool,
        #[case] enable: Option<bool>,
        #[case] expected_enabled: bool,
    ) {
        // Arrange
        let mut config = AppConfig::test(None).unwrap();
        config.service.http.custom.middleware.default_enable = default_enable;
        config
            .service
            .http
            .custom
            .middleware
            .size_limit
            .common
            .enable = enable;

        let context = AppContext::test(Some(config), None, None).unwrap();

        let middleware = RequestBodyLimitMiddleware;

        // Act/Assert
        assert_eq!(middleware.enabled(&context), expected_enabled);
    }

    #[rstest]
    #[case(None, -9970)]
    #[case(Some(1234), 1234)]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn size_limit_priority(#[case] override_priority: Option<i32>, #[case] expected_priority: i32) {
        // Arrange
        let mut config = AppConfig::test(None).unwrap();
        if let Some(priority) = override_priority {
            config
                .service
                .http
                .custom
                .middleware
                .size_limit
                .common
                .priority = priority;
        }

        let context = AppContext::test(Some(config), None, None).unwrap();

        let middleware = RequestBodyLimitMiddleware;

        // Act/Assert
        assert_eq!(middleware.priority(&context), expected_priority);
    }
}
