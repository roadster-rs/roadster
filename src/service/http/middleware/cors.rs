use crate::app::context::AppContext;
use crate::error::RoadsterResult;
use crate::service::http::middleware::Middleware;
use axum::http::{HeaderName, HeaderValue, Method};
use axum::Router;
use axum_core::extract::FromRef;
use itertools::Itertools;
use serde_derive::{Deserialize, Serialize};
use serde_with::{serde_as, skip_serializing_none};
use std::str::FromStr;
use std::time::Duration;
use tower_http::cors::{AllowHeaders, AllowMethods, AllowOrigin, CorsLayer, ExposeHeaders};
use validator::Validate;

#[serde_as]
#[skip_serializing_none]
#[derive(Debug, Clone, Validate, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
#[non_exhaustive]
pub struct CorsConfig {
    #[serde(default)]
    pub preset: CorsPreset,

    /// See <https://docs.rs/tower-http/latest/tower_http/cors/struct.CorsLayer.html#method.allow_credentials>
    #[serde(default)]
    pub allow_credentials: Option<bool>,

    /// See <https://docs.rs/tower-http/latest/tower_http/cors/struct.CorsLayer.html#method.allow_private_network>
    #[serde(default)]
    pub allow_private_network: Option<bool>,

    /// Duration in milliseconds. If a value less than one second (1000 ms) is provided, the
    /// header will not be set by the middleware.
    /// See <https://docs.rs/tower-http/latest/tower_http/cors/struct.CorsLayer.html#method.max_age>
    #[serde(default = "default_max_age")]
    #[serde_as(as = "serde_with::DurationMilliSeconds")]
    pub max_age: Duration,

    /// See <https://docs.rs/tower-http/latest/tower_http/cors/struct.CorsLayer.html#method.allow_headers>
    #[serde(default)]
    pub allow_headers: Option<CorsAllowHeaders>,

    /// See <https://docs.rs/tower-http/latest/tower_http/cors/struct.CorsLayer.html#method.allow_methods>
    #[serde(default)]
    pub allow_methods: Option<CorsAllowMethods>,

    /// See <https://docs.rs/tower-http/latest/tower_http/cors/struct.CorsLayer.html#method.allow_origin>
    #[serde(default)]
    pub allow_origins: Option<CorsAllowOrigins>,

    /// See <https://docs.rs/tower-http/latest/tower_http/cors/struct.CorsLayer.html#method.expose_headers>
    #[serde(default)]
    pub expose_headers: Option<CorsExposeHeaders>,

    /// See <https://docs.rs/tower-http/latest/tower_http/cors/struct.CorsLayer.html#method.vary>
    // Todo: deserialize as HeaderName directly instead of string
    #[serde(default)]
    pub vary: Option<Vec<String>>,
}

fn default_max_age() -> Duration {
    Duration::from_secs(60 * 60)
}

#[derive(Default, Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
#[non_exhaustive]
pub enum CorsPreset {
    /// See <https://docs.rs/tower-http/latest/tower_http/cors/struct.CorsLayer.html#method.new>
    #[default]
    Restrictive,
    /// See <https://docs.rs/tower-http/latest/tower_http/cors/struct.CorsLayer.html#method.permissive>
    Permissive,
    /// See <https://docs.rs/tower-http/latest/tower_http/cors/struct.CorsLayer.html#method.very_permissive>
    VeryPermissive,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "kebab-case")]
#[non_exhaustive]
pub enum CorsAllowHeaders {
    Any,
    MirrorRequest,
    // Todo: deserialize as HeaderName directly instead of string
    List { headers: Vec<String> },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "kebab-case")]
#[non_exhaustive]
pub enum CorsAllowMethods {
    Any,
    MirrorRequest,
    // Todo: deserialize as Method directly instead of string
    Exact { method: String },
    // Todo: deserialize as Method directly instead of string
    List { methods: Vec<String> },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "kebab-case")]
#[non_exhaustive]
pub enum CorsAllowOrigins {
    Any,
    MirrorRequest,
    // Todo: deserialize as HeaderValue directly instead of string
    Exact { origin: String },
    // Todo: deserialize as HeaderValue directly instead of string
    List { origins: Vec<String> },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "kebab-case")]
#[non_exhaustive]
pub enum CorsExposeHeaders {
    Any,
    // Todo: deserialize as HeaderName directly instead of string
    List { headers: Vec<String> },
}

fn parse_header_names(header_names: &[String]) -> RoadsterResult<Vec<HeaderName>> {
    let header_names = header_names
        .iter()
        .map(|header_name| HeaderName::from_str(header_name))
        .try_collect()?;
    Ok(header_names)
}

fn parse_header_values(header_values: &[String]) -> RoadsterResult<Vec<HeaderValue>> {
    let header_values = header_values
        .iter()
        .map(|header_value| HeaderValue::from_str(header_value))
        .try_collect()?;
    Ok(header_values)
}

fn parse_methods(methods: &[String]) -> RoadsterResult<Vec<Method>> {
    let methods = methods
        .iter()
        .map(|method| Method::from_str(method))
        .try_collect()?;
    Ok(methods)
}

pub struct CorsMiddleware;
impl<S> Middleware<S> for CorsMiddleware
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
{
    fn name(&self) -> String {
        "cors".to_string()
    }

    fn enabled(&self, state: &S) -> bool {
        AppContext::from_ref(state)
            .config()
            .service
            .http
            .custom
            .middleware
            .cors
            .common
            .enabled(state)
    }

    fn priority(&self, state: &S) -> i32 {
        AppContext::from_ref(state)
            .config()
            .service
            .http
            .custom
            .middleware
            .cors
            .common
            .priority
    }

    fn install(&self, router: Router, state: &S) -> RoadsterResult<Router> {
        let context = AppContext::from_ref(state);
        let config = &context.config().service.http.custom.middleware.cors.custom;
        let layer = match config.preset {
            CorsPreset::Restrictive => CorsLayer::new(),
            CorsPreset::Permissive => CorsLayer::permissive(),
            CorsPreset::VeryPermissive => CorsLayer::very_permissive(),
        };

        let layer = if config.max_age > Duration::from_secs(1) {
            layer.max_age(config.max_age)
        } else {
            layer
        };

        let layer = config
            .allow_credentials
            .iter()
            .fold(layer, |layer, allow| layer.allow_credentials(*allow));

        let layer = config
            .allow_private_network
            .iter()
            .fold(layer, |layer, allow| layer.allow_private_network(*allow));

        let layer = config.allow_headers.iter().try_fold(
            layer,
            |layer, allow| -> RoadsterResult<CorsLayer> {
                let layer = match allow {
                    CorsAllowHeaders::Any => layer.allow_headers(AllowHeaders::any()),
                    CorsAllowHeaders::MirrorRequest => {
                        layer.allow_headers(AllowHeaders::mirror_request())
                    }
                    CorsAllowHeaders::List { headers } => {
                        layer.allow_headers(AllowHeaders::list(parse_header_names(headers)?))
                    }
                };
                Ok(layer)
            },
        )?;

        let layer = config.expose_headers.iter().try_fold(
            layer,
            |layer, allow| -> RoadsterResult<CorsLayer> {
                let layer = match allow {
                    CorsExposeHeaders::Any => layer.expose_headers(ExposeHeaders::any()),
                    CorsExposeHeaders::List { headers } => {
                        layer.expose_headers(ExposeHeaders::list(parse_header_names(headers)?))
                    }
                };
                Ok(layer)
            },
        )?;

        let layer = config.vary.iter().try_fold(
            layer,
            |layer, header_names| -> RoadsterResult<CorsLayer> {
                let layer = layer.vary(parse_header_names(header_names)?);
                Ok(layer)
            },
        )?;

        let layer = config.allow_origins.iter().try_fold(
            layer,
            |layer, allow| -> RoadsterResult<CorsLayer> {
                let layer = match allow {
                    CorsAllowOrigins::Any => layer.allow_origin(AllowOrigin::any()),
                    CorsAllowOrigins::MirrorRequest => {
                        layer.allow_origin(AllowOrigin::mirror_request())
                    }
                    CorsAllowOrigins::Exact { origin } => {
                        layer.allow_origin(AllowOrigin::exact(HeaderValue::from_str(origin)?))
                    }
                    CorsAllowOrigins::List { origins } => {
                        layer.allow_origin(AllowOrigin::list(parse_header_values(origins)?))
                    }
                };
                Ok(layer)
            },
        )?;

        let layer = config.allow_methods.iter().try_fold(
            layer,
            |layer, allow| -> RoadsterResult<CorsLayer> {
                let layer = match allow {
                    CorsAllowMethods::Any => layer.allow_methods(AllowMethods::any()),
                    CorsAllowMethods::MirrorRequest => {
                        layer.allow_methods(AllowMethods::mirror_request())
                    }
                    CorsAllowMethods::Exact { method } => {
                        layer.allow_methods(AllowMethods::exact(Method::from_str(method)?))
                    }
                    CorsAllowMethods::List { methods } => {
                        layer.allow_methods(AllowMethods::list(parse_methods(methods)?))
                    }
                };
                Ok(layer)
            },
        )?;

        let router = router.layer(layer);

        Ok(router)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::config::AppConfig;
    use crate::testing::snapshot::TestCase;
    use crate::util::serde::Wrapper;
    use insta::assert_toml_snapshot;
    use rstest::{fixture, rstest};

    #[fixture]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn case() -> TestCase {
        Default::default()
    }

    #[rstest]
    #[case(false, Some(true), true)]
    #[case(false, Some(false), false)]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn cors_enabled(
        #[case] default_enable: bool,
        #[case] enable: Option<bool>,
        #[case] expected_enabled: bool,
    ) {
        // Arrange
        let mut config = AppConfig::test(None).unwrap();
        config.service.http.custom.middleware.default_enable = default_enable;
        config.service.http.custom.middleware.cors.common.enable = enable;

        let context = AppContext::test(Some(config), None, None).unwrap();

        let middleware = CorsMiddleware;

        // Act/Assert
        assert_eq!(middleware.enabled(&context), expected_enabled);
    }

    #[rstest]
    #[case(None, -9950)]
    #[case(Some(1234), 1234)]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn cors_priority(#[case] override_priority: Option<i32>, #[case] expected_priority: i32) {
        // Arrange
        let mut config = AppConfig::test(None).unwrap();
        if let Some(priority) = override_priority {
            config.service.http.custom.middleware.cors.common.priority = priority;
        }

        let context = AppContext::test(Some(config), None, None).unwrap();

        let middleware = CorsMiddleware;

        // Act/Assert
        assert_eq!(middleware.priority(&context), expected_priority);
    }

    #[rstest]
    #[case(
        r#"
        [inner]
        type = 'any'
        "#
    )]
    #[case(
        r#"
        [inner]
        type = 'mirror-request'
        "#
    )]
    #[case(
        r#"
        [inner]
        type = 'list'
        headers = ["foo", "bar"]
        "#
    )]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn deserialize_cors_allow_headers(_case: TestCase, #[case] serialized: &str) {
        let value: Wrapper<CorsAllowHeaders> = toml::from_str(serialized).unwrap();
        assert_toml_snapshot!(value);
    }

    #[rstest]
    #[case(
        r#"
        [inner]
        type = 'any'
        "#
    )]
    #[case(
        r#"
        [inner]
        type = 'mirror-request'
        "#
    )]
    #[case(
        r#"
        [inner]
        type = 'exact'
        method = "foo"
        "#
    )]
    #[case(
        r#"
        [inner]
        type = 'list'
        methods = ["foo", "bar"]
        "#
    )]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn deserialize_cors_allow_methods(_case: TestCase, #[case] serialized: &str) {
        let value: Wrapper<CorsAllowMethods> = toml::from_str(serialized).unwrap();
        assert_toml_snapshot!(value);
    }

    #[rstest]
    #[case(
        r#"
        [inner]
        type = 'any'
        "#
    )]
    #[case(
        r#"
        [inner]
        type = 'mirror-request'
        "#
    )]
    #[case(
        r#"
        [inner]
        type = 'exact'
        origin = "foo"
        "#
    )]
    #[case(
        r#"
        [inner]
        type = 'list'
        origins = ["foo", "bar"]
        "#
    )]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn deserialize_cors_allow_origins(_case: TestCase, #[case] serialized: &str) {
        let value: Wrapper<CorsAllowOrigins> = toml::from_str(serialized).unwrap();
        assert_toml_snapshot!(value);
    }

    #[rstest]
    #[case(
        r#"
        [inner]
        type = 'any'
        "#
    )]
    #[case(
        r#"
        [inner]
        type = 'list'
        headers = ["foo", "bar"]
        "#
    )]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn deserialize_cors_expose_headers(_case: TestCase, #[case] serialized: &str) {
        let value: Wrapper<CorsExposeHeaders> = toml::from_str(serialized).unwrap();
        assert_toml_snapshot!(value);
    }
}
