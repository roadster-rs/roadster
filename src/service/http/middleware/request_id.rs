use crate::app_context::AppContext;
use crate::error::RoadsterResult;
use crate::service::http::middleware::Middleware;
use axum::http::HeaderName;
use axum::Router;
use serde_derive::{Deserialize, Serialize};
use std::str::FromStr;
use tower_http::request_id::{MakeRequestUuid, PropagateRequestIdLayer, SetRequestIdLayer};

pub const REQUEST_ID_HEADER_NAME: &str = "request-id";

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case", default)]
pub struct CommonRequestIdConfig {
    pub header_name: String,
}

impl Default for CommonRequestIdConfig {
    fn default() -> Self {
        Self {
            header_name: REQUEST_ID_HEADER_NAME.to_string(),
        }
    }
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case", default)]
pub struct SetRequestIdConfig {
    #[serde(flatten)]
    pub common: CommonRequestIdConfig,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case", default)]
pub struct PropagateRequestIdConfig {
    #[serde(flatten)]
    pub common: CommonRequestIdConfig,
}

pub struct SetRequestIdMiddleware;
impl<S: Send + Sync + 'static> Middleware<S> for SetRequestIdMiddleware {
    fn name(&self) -> String {
        "set-request-id".to_string()
    }

    fn enabled(&self, context: &AppContext<S>) -> bool {
        context
            .config()
            .service
            .http
            .custom
            .middleware
            .set_request_id
            .common
            .enabled(context)
    }

    fn priority(&self, context: &AppContext<S>) -> i32 {
        context
            .config()
            .service
            .http
            .custom
            .middleware
            .set_request_id
            .common
            .priority
    }

    fn install(&self, router: Router, context: &AppContext<S>) -> RoadsterResult<Router> {
        let header_name = &context
            .config()
            .service
            .http
            .custom
            .middleware
            .set_request_id
            .custom
            .common
            .header_name;

        let router = router.layer(SetRequestIdLayer::new(
            HeaderName::from_str(header_name)?,
            MakeRequestUuid,
        ));

        Ok(router)
    }
}

pub struct PropagateRequestIdMiddleware;
impl<S: Send + Sync + 'static> Middleware<S> for PropagateRequestIdMiddleware {
    fn name(&self) -> String {
        "propagate-request-id".to_string()
    }

    fn enabled(&self, context: &AppContext<S>) -> bool {
        context
            .config()
            .service
            .http
            .custom
            .middleware
            .propagate_request_id
            .common
            .enabled(context)
    }

    fn priority(&self, context: &AppContext<S>) -> i32 {
        context
            .config()
            .service
            .http
            .custom
            .middleware
            .propagate_request_id
            .common
            .priority
    }

    fn install(&self, router: Router, context: &AppContext<S>) -> RoadsterResult<Router> {
        let header_name = &context
            .config()
            .service
            .http
            .custom
            .middleware
            .propagate_request_id
            .custom
            .common
            .header_name;

        let router = router.layer(PropagateRequestIdLayer::new(HeaderName::from_str(
            header_name,
        )?));

        Ok(router)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::config::app_config::AppConfig;
    use rstest::rstest;

    #[rstest]
    #[case(false, Some(true), true)]
    #[case(false, Some(false), false)]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn set_request_id_enabled(
        #[case] default_enable: bool,
        #[case] enable: Option<bool>,
        #[case] expected_enabled: bool,
    ) {
        // Arrange
        let mut config = AppConfig::test(None).unwrap();
        config.service.http.custom.middleware.default_enable = default_enable;
        config
            .service
            .http
            .custom
            .middleware
            .set_request_id
            .common
            .enable = enable;

        let context = AppContext::<()>::test(Some(config), None).unwrap();

        let middleware = SetRequestIdMiddleware;

        // Act/Assert
        assert_eq!(middleware.enabled(&context), expected_enabled);
    }

    #[rstest]
    #[case(None, -9990)]
    #[case(Some(1234), 1234)]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn set_request_id_priority(
        #[case] override_priority: Option<i32>,
        #[case] expected_priority: i32,
    ) {
        // Arrange
        let mut config = AppConfig::test(None).unwrap();
        if let Some(priority) = override_priority {
            config
                .service
                .http
                .custom
                .middleware
                .set_request_id
                .common
                .priority = priority;
        }

        let context = AppContext::<()>::test(Some(config), None).unwrap();

        let middleware = SetRequestIdMiddleware;

        // Act/Assert
        assert_eq!(middleware.priority(&context), expected_priority);
    }

    #[rstest]
    #[case(false, Some(true), true)]
    #[case(false, Some(false), false)]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn propagate_request_id_enabled(
        #[case] default_enable: bool,
        #[case] enable: Option<bool>,
        #[case] expected_enabled: bool,
    ) {
        // Arrange
        let mut config = AppConfig::test(None).unwrap();
        config.service.http.custom.middleware.default_enable = default_enable;
        config
            .service
            .http
            .custom
            .middleware
            .propagate_request_id
            .common
            .enable = enable;

        let context = AppContext::<()>::test(Some(config), None).unwrap();

        let middleware = PropagateRequestIdMiddleware;

        // Act/Assert
        assert_eq!(middleware.enabled(&context), expected_enabled);
    }

    #[rstest]
    #[case(None, 9990)]
    #[case(Some(1234), 1234)]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn propagate_request_id_priority(
        #[case] override_priority: Option<i32>,
        #[case] expected_priority: i32,
    ) {
        // Arrange
        let mut config = AppConfig::test(None).unwrap();
        if let Some(priority) = override_priority {
            config
                .service
                .http
                .custom
                .middleware
                .propagate_request_id
                .common
                .priority = priority;
        }

        let context = AppContext::<()>::test(Some(config), None).unwrap();

        let middleware = PropagateRequestIdMiddleware;

        // Act/Assert
        assert_eq!(middleware.priority(&context), expected_priority);
    }
}
