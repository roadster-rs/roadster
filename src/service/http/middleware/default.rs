use crate::app::context::AppContext;
use crate::service::http::middleware::catch_panic::CatchPanicMiddleware;
use crate::service::http::middleware::compression::RequestDecompressionMiddleware;
use crate::service::http::middleware::cors::CorsMiddleware;
use crate::service::http::middleware::request_id::{
    PropagateRequestIdMiddleware, SetRequestIdMiddleware,
};
use crate::service::http::middleware::sensitive_headers::{
    SensitiveRequestHeadersMiddleware, SensitiveResponseHeadersMiddleware,
};
use crate::service::http::middleware::size_limit::RequestBodyLimitMiddleware;
use crate::service::http::middleware::timeout::TimeoutMiddleware;
use crate::service::http::middleware::tracing::req_res_logging::RequestResponseLoggingMiddleware;
use crate::service::http::middleware::tracing::TracingMiddleware;
use crate::service::http::middleware::Middleware;
use axum_core::extract::FromRef;
use std::collections::BTreeMap;

pub fn default_middleware<S>(state: &S) -> BTreeMap<String, Box<dyn Middleware<S>>>
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
{
    let middleware: Vec<Box<dyn Middleware<S>>> = vec![
        Box::new(SensitiveRequestHeadersMiddleware),
        Box::new(SensitiveResponseHeadersMiddleware),
        Box::new(SetRequestIdMiddleware),
        Box::new(PropagateRequestIdMiddleware),
        Box::new(TracingMiddleware),
        Box::new(CatchPanicMiddleware),
        Box::new(RequestDecompressionMiddleware),
        Box::new(TimeoutMiddleware),
        Box::new(RequestBodyLimitMiddleware),
        Box::new(CorsMiddleware),
        Box::new(RequestResponseLoggingMiddleware),
    ];

    middleware
        .into_iter()
        .filter(|middleware| middleware.enabled(state))
        .map(|middleware| (middleware.name(), middleware))
        .collect()
}

#[cfg(test)]
mod tests {
    use crate::app::context::AppContext;
    use crate::config::AppConfig;
    use crate::testing::snapshot::TestCase;
    use insta::assert_toml_snapshot;
    use itertools::Itertools;
    use rstest::{fixture, rstest};

    #[fixture]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn case() -> TestCase {
        Default::default()
    }

    #[rstest]
    #[case(false)]
    #[case(true)]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn default_middleware(_case: TestCase, #[case] default_enable: bool) {
        // Arrange
        let mut config = AppConfig::test(None).unwrap();
        config.service.http.custom.middleware.default_enable = default_enable;

        let context = AppContext::test(Some(config), None, None).unwrap();

        // Act
        let middleware = super::default_middleware(&context);
        let middleware = middleware.keys().collect_vec();

        // Assert
        assert_toml_snapshot!(middleware);
    }
}
