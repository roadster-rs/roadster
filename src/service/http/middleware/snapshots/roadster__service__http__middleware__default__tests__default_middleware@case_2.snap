---
source: src/service/http/middleware/default.rs
expression: middleware
---
[
    'catch-panic',
    'cors',
    'propagate-request-id',
    'request-body-size-limit',
    'request-decompression',
    'sensitive-request-headers',
    'sensitive-response-headers',
    'set-request-id',
    'timeout',
    'tracing',
]
